launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,528 @@
+            item.name = f'{item.name}[shutdown_test]'
+    return items
+
+
+# Adapted from https://github.com/pytest-dev/pytest-asyncio,
+# see their license https://github.com/pytest-dev/pytest-asyncio/blob/master/LICENSE.
+@pytest.mark.tryfirst
+def pytest_pycollect_makeitem(collector, name, obj):
+    """Collect coroutine based launch tests."""
+    if collector.funcnamefilter(name) and is_valid_test_item(obj):
+        item = pytest.Function.from_parent(collector, name=name)
+
+        # Due to how pytest test collection works, module-level pytestmarks
+        # are applied after the collection step. Since this is the collection
+        # step, we look ourselves.
+        transfer_markers(obj, item.cls, item.module)
+        item = pytest.Function.from_parent(collector, name=name)  # To reload keywords.
+
+        if is_launch_test(item):
+            if not is_launch_test_mark_valid(item):
+                # return an item with a warning that's going to be skipped
+                return [item]
+            fixture = get_launch_test_fixture(item)
+            fixturename = fixture.__name__
+            scope = fixture._pytestfixturefunction.scope
+            is_shutdown = has_shutdown_kwarg(item)
+            items = generate_test_items(collector, name, obj, fixturename, is_shutdown, False)
+            if need_shutdown_test_item(obj) and scope != 'function':
+                # for function scope we only need one shutdown item
+                # if not we're going to use two event loops!!!
+                shutdown_items = generate_test_items(collector, name, obj, fixturename, True, True)
+                for item, shutdown_item in zip(items, shutdown_items):
+                    item._launch_testing_shutdown_item = shutdown_item
+                items.extend(shutdown_items)
+            return items
+
+
+def get_error_context_from_obj(obj):
+    """Return formatted information of the object location."""
+    try:
+        fspath = inspect.getsourcefile(obj)
+    except TypeError:
+        return 'location information of the object not found'
+    try:
+        lines, lineno = inspect.getsourcelines(obj)
+    except IOError:
+        return f'file {fspath}: source code not available'
+    error_msg = f'file {fspath}, line {lineno}'
+    for line in lines:
+        line = line.rstrip()
+        error_msg += f'\n  {line}'
+        if line.lstrip().startswith('def'):
+            break
+    return error_msg
+
+
+def is_shutdown_test(item):
+    """Return `True` if the item is a launch test."""
+    return getattr(item, '_launch_testing_is_shutdown', False)
+
+
+def is_same_launch_test_fixture(left_item, right_item):
+    """Return `True` if both items are using the same fixture with the same parameters."""
+    lfn = get_launch_test_fixture(left_item)
+    rfn = get_launch_test_fixture(right_item)
+    if None in (lfn, rfn):
+        return False
+    if lfn != rfn:
```

Comment:
@ivanpauno nit: should we check for identity?  
```suggestion
    if lfn is not rfn:
```